### PR TITLE
Add transaction source for dynamic sampling

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [x] Enhancement
- [x] Refactoring

## :scroll: Description

I've added default `view` source for all the integration, which can be overwritten by the user.

I've picked `view` as its description fitted the most to what it actually is. Example `Home`, `Menu`, `ProfilePage` etc.
https://develop.sentry.dev/sdk/event-payloads/transaction/#transaction-annotations

## :bulb: Motivation and Context

At the moment the source is always `custom`, because we first create empty transaction and then update it with the correct route data, which causes the source to be set as `custom`.

https://github.com/getsentry/team-mobile/issues/46

## :green_heart: How did you test it?

🚨 To be done when sentry backend will allow dynamic sampling for React Native.

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Comment by marandaneto** [comment]
@krystofwoldrich `view` looks fine, Android used `component`
https://github.com/getsentry/sentry-java/blob/0d55590e368f954db1a06777867f4941be424889/sentry-android-core/src/main/java/io/sentry/android/core/ActivityLifecycleIntegration.java#L192
Can you and @adinauer figure this out? They don't need to be the same but I'd expect them to be, at the end, they are both screen's names most likely, the difference is that on Android its an Activity, and on RN is a routing that ends up being the screen's name as well.
Maybe @lforst can chime in as well since there are integrations on the FE as well, I'd expect to align the React integration on the Web.

**Comment by lforst** [comment]
In browser land we almost exclusively use route/url.

I originally thought that view was the most fitting here judging by the name. But if we use component in android I'd use that just to be consistent. If we want we could update the docs to be a little more specific on what the differences are.

**Comment by adinauer** [comment]
Seems android should also be `view` in some places where `component` is used at the moment. We can also open an issue on the Java SDK to change it.

File: src/js/tracing/reactnativenavigation.ts [hunk 1]
```diff
@@ -171,29 +171,16 @@ export class ReactNativeNavigationInstrumentation extends InternalRoutingInstrum
...
 
-        if (finalContext.sampled === false) {
-          logger.log(
-            `[${ReactNativeNavigationInstrumentation.name}] Will not send transaction "${finalContext.name}" due to beforeNavigate.`
-          );
-        }
+        const isCustomName = updatedContext.name !== finalContext.name;
+        this._latestTransaction.setName(
+          finalContext.name,
+          isCustomName ? 'custom' : defaultTransactionSource,
```
Comment (marandaneto) [inline src/js/tracing/reactnativenavigation.ts:180]: `custom` can also be a const

File: test/tracing/reactnavigationv4.test.ts [hunk 2]
```diff
@@ -498,6 +501,7 @@ describe('ReactNavigationV4Instrumentation', () => {
       if (confirmedContext) {
         expect(confirmedContext.name).toBe(route2.routeName);
         expect(confirmedContext.data).toBeDefined();
+        expect(confirmedContext.metadata).toBeUndefined();
```
Comment (marandaneto) [inline test/tracing/reactnavigationv4.test.ts:504]: I'd remove this line and the others but the `onRouteConfirmed`.

File: src/js/tracing/utils.ts [hunk 3]
```diff
@@ -1,7 +1,9 @@
 import { IdleTransaction, Span, Transaction } from '@sentry/tracing';
-import { TransactionContext } from '@sentry/types';
+import { TransactionContext, TransactionSource } from '@sentry/types';
 import { timestampInSeconds } from '@sentry/utils';
 
+export const defaultTransactionSource: TransactionSource = 'view';
```
Comment (marandaneto) [inline src/js/tracing/utils.ts:5]: Let's change to `component`.

**Review by marandaneto — APPROVED** [review]

**Comment by marandaneto** [comment]
@krystofwoldrich Thank you for doing that, I liked the refactoring, it makes it easier to read.

**Comment by marandaneto** [comment]
@krystofwoldrich after merging and releasing this, let's update the [docs](https://docs.sentry.io/product/data-management-settings/server-side-sampling/getting-started/#current-limitations).

File: test/tracing/reactnavigationv4.test.ts (hunk 2 above)
Comment (krystofwoldrich) [reply test/tracing/reactnavigationv4.test.ts:504]: This one is actually on `onRouteConfirmed` too, just the GitHub UI makes it look like it is in `transaction not attached on a cancelled navigation`.

**Comment by krystofwoldrich** [comment]
Docs update:
- https://github.com/getsentry/sentry-docs/pull/5486

**Comment by krystofwoldrich** [comment]
The only missing part now is test with backend.

That can happen when the React Native will be allowed for Dynamic Sampling.

**Comment by krystofwoldrich** [comment]
Test on production, source is send correctly. Dynamic sampling on server is also working. From 10 transaction only 2 were processed when sampling set to 20%.

**Pull request merged**